File: src/main/java/com/extendedae_plus/common/wireless/LinkMaster.java
package com.extendedae_plus.common.wireless;

import com.extendedae_plus.common.wireless.linkApi.ILinkHost;
import com.extendedae_plus.common.wireless.linkApi.Label;
import com.extendedae_plus.common.wireless.linkApi.RegistryLink;
import net.minecraft.server.level.ServerLevel;
import org.jetbrains.annotations.Nullable;

import java.util.UUID;

/**
 * 主收发器端逻辑：负责在频率变化/加载时向注册中心登记唯一主端，卸载时反注册。
 * 方块实体应在合适的生命周期中调用 register/unregister。
 */
public class LinkMaster {
    private final ILinkHost host;
    private boolean registered;

    public LinkMaster(ILinkHost host) {
        this.host = host;
    }

    public Label getLabel() {
        return this.host.getLabel();
    }

    public void updateInfo(Label label, @Nullable UUID placer) {
        if (this.host.getLabel().equals(label)
                && this.host.getPlacer() == placer) return;

        this.unregister();
        this.host.setLabel(label);
        this.host.setPlacer(placer);
        this.register();
    }

    public void setPlacer(@Nullable UUID placer) {
        this.host.setPlacer(placer);
    }

    public void setLabel(Label label) {
        if (this.host.getLabel().equals(label)) return;

        this.unregister();
        this.host.setLabel(label);
        this.register();
    }

    public boolean register() {
        ServerLevel level = this.host.getServerLevel();
        if (level == null || this.host.getLabel().data.isEmpty()) return false;
        boolean succeed = RegistryLink.registerMaster(this.host);
        this.registered = succeed;
        return succeed;
    }

    public void unregister() {
        ServerLevel level = this.host.getServerLevel();
        if (!this.registered || level == null || this.host.getLabel().data.isEmpty()) return;
        RegistryLink.unregisterMaster(this.host);
        this.registered = false;
    }

    public boolean connected() {
        return RegistryLink.countListener(this.host.getLabel()) > 0;
    }

    public void onUnloadOrRemove() {
        unregister();
    }
}

File: src/main/kotlin/com/extendedae_plus/common/wireless/LinkMaster.kt
package com.extendedae_plus.common.wireless

import com.extendedae_plus.common.wireless.linkApi.ILinkHost
import com.extendedae_plus.common.wireless.linkApi.Label
import com.extendedae_plus.common.wireless.linkApi.RegistryLink
import java.util.*

/**
 * 主收发器端逻辑：负责在频率变化/加载时向注册中心登记唯一主端，卸载时反注册。
 * 方块实体应在合适的生命周期中调用 register/unregister。
 */
class LinkMaster(private val host: ILinkHost) {
    private var registered = false

    var label: Label
        get() = this.host.label
        set(label) {
            if (this.host.label == label) return

            this.unregister()
            this.host.label = label
            this.register()
        }

    fun updateInfo(label: Label, placer: UUID?) {
        if (this.host.label == label
            && this.host.placer == placer
        ) return

        this.unregister()
        this.host.label = label
        this.host.placer = placer
        this.register()
    }

    fun setPlacer(placer: UUID?) {
        this.host.placer = placer
    }

    fun register(): Boolean {
        val level = this.host.serverLevel
        if (level == null || this.host.label.data.isEmpty) return false
        val succeed = RegistryLink.registerMaster(this.host)
        this.registered = succeed
        return succeed
    }

    fun unregister() {
        val level = this.host.serverLevel
        if (!this.registered || level == null || this.host.label.data.isEmpty) return
        RegistryLink.unregisterMaster(this.host)
        this.registered = false
    }

    fun connected() = RegistryLink.countListener(this.host.label) > 0

    fun onUnloadOrRemove() = this.unregister()
}
